lib/cosmos/interfaces/interface.rb

```diff
@@ -218,62 +251,100 @@ def set_option(option_name, option_values)
-      @adapter_params[adapter] = params.clone
+    def set_adapter_params(procotol, params)
+      @adapter_params[procotol] = params.clone
     end
 
-    # Called to perform modifications on a read packet before it is inserted
-    # into the current value table.
+    # Called to read data and manipulate it until enough data is
+    # returned. The definition of 'enough data' changes depending on the
+    # protocol used which is why this method exists. This method is also used
+    # to perform operations on the data before it can be interpreted as packet
+    # data such as decryption. After this method is called the post_read_data
+    # method is called.
     #
-    # @param packet [Packet] Original packet
-    # @return [Packet] Potentially modified packet
-    def post_read_packet(packet)
-      packet
+    # @return [String] Raw packet data
+    def read_data
     end
 
-    # Called to perform modifications on read data before making it into a packet
+    # Called to perform modifications on read data before making it into a packet.
+    # TODO: Example of using this method
+    # After this method is called the post_read_packet method is called.
     #
     # @param packet_data [String] Raw packet data
     # @return [String] Potentially modified packet data
     def post_read_data(packet_data)
       packet_data
     end
 
-    # Called to perform modifications on a command packet before it is send
+    # Called to perform modifications on a read packet before it is identified
+    # and inserted into the current value table. This is the final place to
+    # modify data before it is used by the COSMOS system. An example would be
+    # to set the packet timestamp. After this method is called the
+    # post_identify_packet method is called.
+    #
+    # @param packet [Packet] Original packet
+    # @return [Packet] Potentially modified packet
+    def post_read_packet(packet)
+      packet
+    end
+
+    # This method is called by the CmdTlmServer after each read packet is
+    # identified. It can be used to perform custom processing/monitoring as
+    # each packet is received by the CmdTlmServer.
+    #
+    # @param packet [Packet] The identified packet read from the interface
+    def post_identify_packet(packet)
+    end
+
+    # Called to perform modifications on a command packet before it is turned
+    # into packet data to send. An example would be to override a known packet
+    # value. After this method is called the pre_write_data method is called.
     #
     # @param packet [Packet] Original packet
     # @return [Packet] Potentially modified packet
     def pre_write_packet(packet)
       packet
     end
 
-    # Called to perform modifications on write data before making it into a packet
+    # Called to write a packet to the underlying interface
+    #
+    # @param packet [Packet] Packet data
+    # @return [Packet] The original packet
+    def write_packet(packet)
+      packet
```

Comment:
By default we don't do anything and just return the packet and don't write anything.